python/ray/data/_internal/execution/streaming_executor_state.py

```diff
@@ -330,6 +343,44 @@ def select_operator_to_run(
     )
 
 
+def _try_to_scale_up_cluster(topology: Topology, execution_id: str):
+    """Try to scale up the cluster to accomodate the provided in-progress workload.
+
+    This makes a resource request to Ray's autoscaler consisting of the current,
+    aggregate usage of all operators in the DAG + the incremental usage of all operators
+    that are ready for dispatch (i.e. that have inputs queued). If the autoscaler were
+    to grant this resource request, it would allow us to dispatch one task for every
+    ready operator.
+
+    Note that this resource request does not take the global resource limits or the
+    liveness policy into account; it only tries to make the existing resource usage +
+    one more task per ready operator feasible in the cluster.
+
+    Args:
+        topology: The execution state of the in-progress workload for which we wish to
+            request more resources.
+    """
+    # Get resource usage for all ready ops.
+    usage = ExecutionResources()
+    for op, state in topology.items():
+        usage = usage.add(op.current_resource_usage())
+        # Only include incremental resource usage for ops that are ready for
+        # dispatch.
+        if state.num_queued() > 0:
+            # TODO(Clark): Scale up more aggressively by adding incremental resource
+            # usage for more than one bundle in the queue for this op?
+            usage = usage.add(op.incremental_resource_usage())
+    resource_request = {}
+    # Round usage to ints, since autoscaler only accepts integral resource requests.
+    if usage.cpu:
+        resource_request["CPU"] = math.ceil(usage.cpu)
+    if usage.gpu:
+        resource_request["GPU"] = math.ceil(usage.gpu)
+    # Make autoscaler resource request.
+    actor = get_or_create_autoscaling_requester_actor()
+    actor.request_resources.remote(resource_request, execution_id)
```

Comment:
Should we also throttle the calls to this to e.g., at most once every 5 seconds? Actor API operations are somewhat expensive and we shouldn't e.g., call it thousands of times a second from streaming executor.

Reply:
This happens only when resource crunching, it seems not to have massive requests.

Reply:
This could still be hundreds of times a second though right? We should be careful here and avoid possible overloads.

Reply:
Added.